src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1888 @@
+        # candidate for perpendicular line is in 3rd coordinate.
+
+        # If this is a self common line with respect to x-equator then the actual self
+        # common line(s) is given by the self relative rotations given by the y and z
+        # rotation (by 180 degrees) group members, i.e. Ri^TgyRj and Ri^TgzRj
+        scl_angles[eq_class == 1] = scl_angles[eq_class == 1][:, :, [1, 2, 0]]
+        scl_angles[eq_class == 1, :, 0] = scl_angles[eq_class == 1][:, :, 0, [1, 0]]
+
+        # If this is a self common line with respect to y-equator then the actual self
+        # common line(s) is given by the self relative rotations given by the x and z
+        # rotation (by 180 degrees) group members, i.e. Ri^TgxRj and Ri^TgzRj
+        scl_angles[eq_class == 2] = scl_angles[eq_class == 2][:, :, [0, 2, 1]]
+        scl_angles[eq_class == 2, :, 0] = scl_angles[eq_class == 2][:, :, 0, [1, 0]]
+
+        # If this is a self common line with respect to z-equator then the actual self
+        # common line(s) is given by the self relative rotations given by the x and y
+        # rotation (by 180 degrees) group members, i.e. Ri^TgxRj and Ri^TgyRj
+        # No need to rearrange entries, the "real" common lines are already in
+        # indices 1 and 2, but flip one common line to antipodal.
+        scl_angles[eq_class == 3, :, 0] = scl_angles[eq_class == 3][:, :, 0, [1, 0]]
+
+        # Make sure angle range is < 180 degrees.
+        # p1 marks "equator" self-commonlines where both entries of the first
+        # scl are greater than both entries of the second scl.
+        p1 = scl_angles[eq_class > 0, :, 0] > scl_angles[eq_class > 0, :, 1]
+        p1 = p1[:, :, 0] & p1[:, :, 1]
+        # p2 marks "equator" self-commonlines where the angle range between the
+        # first and second sets of self-commonlines is greater than 180.
+        p2 = scl_angles[eq_class > 0, :, 0] - scl_angles[eq_class > 0, :, 1] < -np.pi
+        p2 = p2[:, :, 0] | p2[:, :, 1]
+        p = p1 | p2
+
+        # Swap entries satisfying either of the above conditions.
+        scl_angles[eq_class > 0] = (
+            scl_angles[eq_class > 0][:, :, [1, 0, 2]] * p[:, :, None, None]
+            + scl_angles[eq_class > 0] * ~p[:, :, None, None]
+        )
+
+        # Convert from angles [0,2*pi) to degrees [0, 360).
+        return np.round(scl_angles * 180 / np.pi) % 360
+
+    def _generate_scl_indices(self, scl_angles, eq_class):
+        """
+        Generate self-commonline indices. This includes a set of linear indices for
+        all candidate rotations as well as lists of self-commonline index ranges for
+        equator candidates.
+
+        :param scl_angles: Self-commonline angles, shape (n_sphere_grid, n_inplane_rots, 3, 2).
+        :param eq_class: Equator classification for the sphere_grid points represented
+            by the first axis of `scl_angles`.
+
+        :returns:
+            - scl_indices, self-commonline linear indices.
```

Comment:
No, not a new sphinx style. It's just markdown. It renders nicely in the docs. There is not an official way to document multiple returns in sphinx since technically you never have multiple returns, just a tuple return.

Happy to change this to a single return if you think that would be better.

Reply:
Cool, I will keep in mind for other prs, thanks! I didn't take this pr to the point of generating the docs due to the changes, so I didn't really know if it worked or not. Thanks